plugins/optimization-detective/detect.js

```diff
@@ -364,43 +377,107 @@ export default async function detect( {
+			return;
+		}
 
-	// Yield to main before sending data to server to further break up task.
-	await new Promise( ( resolve ) => {
-		setTimeout( resolve, 0 );
-	} );
+		for ( const [ webVital, metric ] of Object.entries(
+			webVitalsMetrics
+		) ) {
+			if ( ! metric.length ) {
+				continue;
+			}
 
-	try {
-		const response = await fetch( restApiEndpoint, {
-			method: 'POST',
-			headers: {
-				'Content-Type': 'application/json',
-				'X-WP-Nonce': restApiNonce,
-			},
-			body: JSON.stringify( urlMetrics ),
-		} );
+			urlMetrics.webVitals[ webVital ] = metric[ 0 ].value;
+		}
+
+		const lcpMetric = webVitalsMetrics.LCP.at( -1 );
 
-		if ( response.status === 200 ) {
-			setStorageLock( getCurrentTime() );
+		for ( const i in urlMetrics.elements ) {
+			const elementMetrics = urlMetrics.elements[ i ];
+
+			elementMetrics.isLCP =
+				elementMetrics.target === lcpMetric?.entries[ 0 ]?.element;
+			elementMetrics.isLCPCandidate = webVitalsMetrics.LCP.some(
+				( lcpMetricCandidate ) =>
+					lcpMetricCandidate.entries[ 0 ]?.element ===
+					elementMetrics.target
+			);
 		}
 
+		// TODO: Collect Server-Timing information.
+
+		// TODO: Detect form factor à la CRuX.
+		// See https://developer.chrome.com/docs/crux/methodology/dimensions
+
 		if ( isDebug ) {
-			const body = await response.json();
+			log( 'Current URL metrics:', urlMetrics );
+		}
+
+		try {
+			const response = await fetch( restApiEndpoint, {
+				method: 'POST',
+				headers: {
+					'Content-Type': 'application/json',
+					'X-WP-Nonce': restApiNonce,
+				},
+				body: JSON.stringify( urlMetrics ),
+				keepalive: true,
+			} );
+
 			if ( response.status === 200 ) {
-				log( 'Response:', body );
-			} else {
-				error( 'Failure:', body );
+				setStorageLock( getCurrentTime() );
+			}
+
+			if ( isDebug ) {
+				const body = await response.json();
+				if ( response.status === 200 ) {
+					log( 'Response:', body );
+				} else {
+					error( 'Failure:', body );
+				}
+			}
+		} catch ( err ) {
+			if ( isDebug ) {
+				error( err );
 			}
 		}
-	} catch ( err ) {
-		if ( isDebug ) {
-			error( err );
-		}
+
+		// Clean up.
+		breadcrumbedElementsMap.clear();
+
+		webVitalsMetrics.LCP = [];
+		webVitalsMetrics.CLS = [];
+		webVitalsMetrics.INP = [];
+		webVitalsMetrics.TTFB = [];
 	}
 
-	// Clean up.
-	breadcrumbedElementsMap.clear();
+	log( 'Adding event listeners for backgrounding and unloading.' );
+
+	await new Promise( ( resolve ) => {
+		setTimeout( resolve, 1000 );
+	} );
+
+	await sendData();
+
+	// Report all available metrics whenever the page is backgrounded or unloaded.
+	addEventListener( 'visibilitychange', () => {
+		if ( document.visibilityState === 'hidden' ) {
+			sendData();
+		}
+	} );
+
+	// NOTE: Safari does not reliably fire the `visibilitychange` event when the
+	// page is being unloaded. If Safari support is needed, you should also flush
+	// the queue in the `pagehide` event.
+	addEventListener( 'pagehide', () => {
+		sendData();
+	} );
```

Comment:
I've largely adopted this in https://github.com/WordPress/performance/pull/1373 now so that instead of sending the gathered URL metric as soon as the page has loaded and is idle, it instead sends it via `navigator.sendBeacon()` when the page is left. It just sends it once per page. The intention is that client-side extensions can hook in to finalize the URL metric object before it are sent, as can be seen in that PR currently for using a `ResizeObserver` to update the ultimate height of embeds rather than just the initial height.